nixos/lib/utils.nix

```diff
@@ -45,6 +45,23 @@ rec {
    replaceChars ["/" "-" " "] ["-" "\\x2d" "\\x20"]
    (removePrefix "/" s);
 
+  # Quotes an argument for use in Exec* service lines.
+  # systemd accepts "-quoted strings with escape sequences, toJSON produces
+  # a subset of these.
+  # Additionally we escape % to disallow expansion of % specifiers. Any lone ;
+  # in the input will be turned it ";" and thus lose its special meaning.
+  # Every $ is escaped to $$, this makes it unnecessary to disable environment
+  # substitution for the directive.
+  escapeSystemdExecArg = arg:
+    let
+      s = builtins.toJSON (toString arg);
+    in
+      replaceChars [ "%" "$" ] [ "%%" "$$" ] s;
+
+  # Quotes a list of arguments into a single string for use in a Exec*
+  # line. Environment substitution may still apply.
```

Comment:
the rest of the line may still do it, the results of the escaping will not. this could have been more clear. how about
```
Quotes a list of arguments into a single string for use in a Exec*
line. Environment substitution will not be applied in escaped arguments,
but may still be used in the remainder of the directive.
```

Reply:
Ok, this would have been obvious to me, but maybe someone needs to be reminded. Not sure; you decide. May become redundant with #154123 anyway.

Reply:
we've removed the whole thing about substitutions here. probably ends up being more confusing than helpful anyway.